### PR TITLE
cl: fix Gloas bid selection, validation and publication

### PR DESCRIPTION
**Pull request by yperbasis**
## Summary

Gloas `ExecutionPayloadBid.value` is denominated in Gwei, while execution-layer values and block-production metadata use Wei. Production compared the raw Gwei bid with the local Wei value, so external bids almost never won. It also narrowed execution values to `uint64`, which corrupted metadata above the Wei limit of that type.

Once external bid selection became reachable, a bid validated when it entered the pool could still become invalid against the production state. Parent execution requests can change builder eligibility before the current bid is processed. Selecting such a bid made the final state transition fail even though a valid self-built block was available. Gloas selection also retained local blob data after choosing an external bid and made a legacy Builder API request whose result could not be used.

Consensus block value metadata was emitted in Gwei, while the beacon API requires Wei.

## Changes

- Keep execution values as exact `big.Int` Wei values through selection and response metadata.
- Convert external Gloas bid values from Gwei to Wei before comparison and reporting.
- Keep `builder_boost_factor` scoped to the legacy Builder API path; compare Gloas ePBS bids directly after converting both values to Wei.
- Run the complete canonical `ProcessBlock` transition for a winning external bid on copies of the production state and self-built block.
- Reuse the successful candidate state and reward collector instead of replaying the transition.
- Keep the original self-built block and state untouched while evaluating the external candidate, then process them if its transition fails.
- Evict only deterministically invalid external bids, including when the self-build fallback also fails, while preserving any newer concurrently stored bid.
- Clear local blobs and KZG proofs when an external bid wins.
- Build local Gloas data-column sidecars only for self-built bids; external builders publish their payload data.
- Keep cached self-build envelopes aligned with the produced block by retaining a same-slot envelope only when it commits to the same beacon block.
- Skip the legacy Builder API `getHeader` request on Gloas, where external bids arrive through ePBS gossip.
- Reject empty, non-decimal (including signed), and out-of-range Builder API block values before weighted comparison.
- Convert consensus block value metadata from Gwei to Wei with overflow-safe arithmetic.
- Log sidecars actually built by the node and avoid unused signed-header merkleization during Gloas publication.

Concurrent same-slot/DVT envelope caching requires a multi-entry design and is tracked in #23781.

The Gwei denomination follows the [Gloas `ExecutionPayloadBid` specification](https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/beacon-chain.md#executionpayloadbid).

### CODE REVIEW

File: cl/beacon/handler/block_production.go [hunk 1]
```diff
@@ -1188,9 +1208,9 @@ func (a *ApiHandler) produceBeaconBody(
 		}
 		// Determine block value
 		if blockValue == nil {
-			executionValue = 0
+			executionValue.SetUint64(0)
 		} else {
-			executionValue = blockValue.Uint64()
+			executionValue.Set(blockValue)
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply cl/beacon/handler/block_production.go:1350]: Addressed in 4835a9503e. `TestProduceBlockPreservesLargeExecutionValue` now feeds `2^64` from the mocked execution layer through `produceBlock` and asserts the exact `ExecutionValue`. I also verified that the test fails with the former `Uint64` narrowing and passes with the `big.Int` path.

File: cl/beacon/handler/block_production.go [hunk 2]
```diff
@@ -984,6 +1087,9 @@ func (a *ApiHandler) getBuilderPayload(
 	if !strings.EqualFold(header.Version, curVersion) {
 		return nil, fmt.Errorf("invalid version %s, expected %s", header.Version, curVersion)
 	}
+	if header.BlockValue() == nil {
+		return nil, fmt.Errorf("invalid builder block value %q", header.Data.Message.Value)
+	}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply cl/beacon/handler/block_production.go]: Fixed in b4c820317e. ExecutionHeader.BlockValue now rejects negative parsed values at the Builder API parsing boundary, so getBuilderPayload cannot accept "-1" as an unsigned block value. I also added "-1" to TestGetBuilderPayloadRejectsInvalidBlockValue; the targeted test and the builder/handler package tests pass.

File: cl/beacon/builder/types.go [hunk 3]
```diff
@@ -48,10 +48,14 @@ func (h ExecutionHeader) BlockValue() *big.Int {
...
-	//blockValue := binary.LittleEndian.Uint64([]byte(h.Data.Message.Value))
 	blockValue, ok := new(big.Int).SetString(h.Data.Message.Value, 10)
 	if !ok {
 		log.Warn("cannot parse block value", "value", h.Data.Message.Value)
+		return nil
+	}
+	if blockValue.Sign() < 0 {
+		log.Warn("negative block value", "value", h.Data.Message.Value)
+		return nil
 	}
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply cl/beacon/builder/types.go]: Fixed in 14b904dfbe. ExecutionHeader.BlockValue now rejects values outside the uint256 range, including values with a bit length above 256. TestGetBuilderPayloadRejectsInvalidBlockValue now covers 2^256 as the over_uint256 boundary case. The focused tests, make test-short, repeated lint, and the required builds pass.

**Review by AskAlexSharov — APPROVED** [review]

File: cl/beacon/handler/block_production.go [hunk 4]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+	selfBuildMachine, selfBuildErr := processBlockForProduction(baseState, block)
+	if selfBuildErr != nil {
+		return baseState, selfBuildMachine, fmt.Errorf(
+			"external builder transition failed: %w; self-build fallback failed: %w",
+			candidateErr,
+			selfBuildErr,
+		)
+	}
+
+	removed := a.epbsPool.RemoveHighestBid(bidKey, externalBid)
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go]: Eviction is not limited to bid defects. `candidateErr` can come from anything `ProcessBlock` returns — e.g. `processExecutionPayloadBid: failed to verify bid signature: <bls backend error>`, `GetBeaconProposerIndex` failure, or an epoch-accessor error — not only from a bid that the parent execution requests invalidated.

The self-build re-run does not separate the two cases: a self-build bid takes the `BuilderIndexSelfBuild` branch in `ProcessExecutionPayloadBid`, which skips the builder-activity, coverage and signature checks entirely. So "the same base state accepts the self-build" only proves the rest of the block is fine; it says nothing about whether the external bid was really invalid. A transient failure therefore permanently drops a valid highest bid for the slot.

Consider evicting only for the error classes that are bid-specific (wrap them with a sentinel in `ProcessExecutionPayloadBid`, or match on the operation that failed).

File: cl/beacon/handler/block_production.go [hunk 5]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+		blockMachine, err := processBlockForProduction(baseState, block)
+		return baseState, blockMachine, err
+	}
+	bidKey := pool.HighestBidKey{
+		Slot:            block.Slot,
+		ParentBlockHash: selfBid.Message.ParentBlockHash,
+		ParentBlockRoot: selfBid.Message.ParentBlockRoot,
+	}
+	externalBid, found := a.epbsPool.HighestBids.Get(bidKey)
+	selectedValueWei, selected := selectHigherGloasBidValue(block.ExecutionValue, externalBid, boostFactor)
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go]: Applying the produceBlockV3 `builder_boost_factor` query parameter to Gloas ePBS bids couples two different knobs.

That parameter is documented for the relay/blinded flow ("choosing between a builder payload header and payload from the paired execution node"). Validator clients set it to `0` to mean "this validator has no relay" — Lighthouse does exactly that when builder proposals are off. Under this change that value now also switches off protocol-level p2p bid selection, which does not involve a relay at all.

beacon-APIs puts the Gloas boost factor somewhere else: `apis/validator/block.v4.yaml` says `builder_boost_factor` arrives in the produceBlockV4 `BuilderConfig` request body and applies to p2p bids. Reusing the V3 query param is a silent behaviour change for operators who run Gloas without a relay.

File: cl/beacon/handler/block_production.go [hunk 6]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
 	return block, nil
 }
 
+func shouldRequestBuilderHeader(stateVersion clparams.StateVersion, builderEnabled, clientAvailable bool) bool {
+	// Gloas receives external builder bids through ePBS gossip instead of the legacy Builder API.
+	return stateVersion.Before(clparams.GloasVersion) && builderEnabled && clientAvailable
+}
+
+func preferLocalExecutionValue(localValueWei, builderValueWei *big.Int, boostFactor uint64) bool {
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go:954]: Nil `localValueWei` panics here: `new(big.Int).Mul(nil, big.NewInt(100))` dereferences `x.abs`.

`processProducedBlock` feeds this the raw field `block.ExecutionValue`, not `block.GetExecutionValue()`. `BlindOrExecutionBeaconBlock` ships a `GetExecutionValue()` accessor precisely because the field is allowed to be nil, and every other read in this file goes through it. Today `produceBlock` always sets the field, so the panic is not reachable from the handler — but the guard is one refactor away from being needed, and the asymmetry invites it.

Use `block.GetExecutionValue()` at the call site, or add a nil check next to the existing `builderValueWei == nil` one.

File: cl/pool/epbs_pool.go [hunk 7]
```diff
@@ -50,6 +54,24 @@ type EpbsPool struct {
...
 
+// StoreHighestBid replaces the current entry for key.
+func (p *EpbsPool) StoreHighestBid(key HighestBidKey, bid *cltypes.SignedExecutionPayloadBid) {
+	p.highestBidsMu.Lock()
+	defer p.highestBidsMu.Unlock()
+	p.HighestBids.Add(key, bid)
+}
+
+// RemoveHighestBid preserves a concurrently stored replacement for the same key.
+func (p *EpbsPool) RemoveHighestBid(key HighestBidKey, bid *cltypes.SignedExecutionPayloadBid) bool {
```
Comment (AskAlexSharov) [inline cl/pool/epbs_pool.go:72]: Eviction can forfeit external bidding for the whole slot.

`HighestBids` keeps one entry per key, and `executionPayloadBidService.validateHighestBid` drops every bid that is not strictly higher, so lower bids from other builders were never stored. After `RemoveHighestBid` there is nothing left to fall back to. The same builder cannot re-bid either: `storeValidBid` adds `(builderIndex, slot)` to `seenCache` before storing, and that entry survives the eviction, so a corrected bid from that builder is rejected with `ErrIgnore`.

Net effect: one bad top bid disables ePBS external bidding for that slot unless a different builder happens to gossip afterwards. Worth either keeping a runner-up, or clearing the matching `seenCache` entry when a bid is evicted.

File: cl/beacon/handler/block_production.go [hunk 8]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+	selfBlobs := block.Blobs
+	selfKzgProofs := block.KzgProofs
+	selfExecutionValue := block.ExecutionValue
+	block.BeaconBody.SignedExecutionPayloadBid = externalBid
+	block.Blobs = nil
+	block.KzgProofs = nil
+	block.ExecutionValue = selectedValueWei
+
+	candidateMachine, candidateErr := processBlockForProduction(candidateState, block)
+	if candidateErr == nil {
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go:1043]: When the external bid wins, a self-build envelope cached for the same slot by an earlier call is left behind.

`GetEthV3ValidatorBlock` does `a.selfBuildEnvelopes.Add(targetSlot, envelope)` on the self-build path. Validator clients retry block production for a slot (retries, DVT, multi-BN), so a first call can self-build and cache, and a second call for the same slot can select the external bid. `GET /eth/v1/validator/execution_payload_envelopes/{slot}` then still serves the discarded local envelope, and the by-builder-index variant only guards against a mismatched `builder_index`.

Removing the slot entry from `selfBuildEnvelopes` when an external bid is selected would close this.

File: cl/beacon/handler/block_production.go [hunk 9]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+}
+
+func selectHigherGloasBidValue(
+	localValueWei *big.Int,
+	externalBid *cltypes.SignedExecutionPayloadBid,
+	boostFactor uint64,
+) (*big.Int, bool) {
+	if externalBid == nil || externalBid.Message == nil {
+		return localValueWei, false
+	}
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go:969]: The bid total ignores `execution_payment`.

beacon-APIs `apis/validator/block.v4.yaml` defines the quantity that competes with the local value as the bid total payment: "A bid's total payment counts its `execution_payment` at most at the `max_execution_payment` that applies to it". This helper uses `Message.Value` only.

It is inert today because `executionPayloadBidService` rejects any bid with `ExecutionPayment != 0`, but this function is where the rule has to land once that restriction is lifted, and the Gwei→Wei conversion added here will have to cover it too. A short note or a TODO with an issue would keep the two from drifting.

File: cl/beacon/handler/block_production.go [hunk 10]
```diff
@@ -843,7 +841,7 @@ func (a *ApiHandler) produceBlock(
 		defer func() {
 			a.logger.Debug("MevBoost", "slot", targetSlot, "duration", time.Since(start))
 		}()
-		if a.routerCfg.Builder && a.builderClient != nil {
+		if shouldRequestBuilderHeader(stateVersion, a.routerCfg.Builder, a.builderClient != nil) {
 			builderHeader, builderErr = a.getBuilderPayload(ctx, baseState, targetSlot)
 			if builderErr != nil && !errors.Is(builderErr, errBuilderNotEnabled) {
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go:876]: This `errBuilderNotEnabled` branch is now dead.

`shouldRequestBuilderHeader` already requires `builderEnabled && clientAvailable`, which is the exact negation of the `!a.routerCfg.Builder || a.builderClient == nil` guard that makes `getBuilderPayload` return `errBuilderNotEnabled`. So inside this `if`, that error can never come back and the `errors.Is` check never fires.

Either drop the `errors.Is` special case here (and the sentinel, if `getBuilderPayload` has no other caller), or drop the two bool parameters from `shouldRequestBuilderHeader` and keep the sentinel as the single source of truth.

File: cl/beacon/handler/block_production.go [hunk 11]
```diff
@@ -984,6 +1090,9 @@ func (a *ApiHandler) getBuilderPayload(
 	if !strings.EqualFold(header.Version, curVersion) {
 		return nil, fmt.Errorf("invalid version %s, expected %s", header.Version, curVersion)
 	}
+	if header.BlockValue() == nil {
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go]: The builder value string gets parsed twice.

`BlockValue()` runs `big.Int.SetString` plus the range checks here, and `produceBlock` calls `builderHeader.BlockValue()` again a few lines later to do the comparison. Both parse the same decimal string.

Returning the parsed value from the validation (or caching it on the header) removes the second parse and, more usefully, removes the chance of the two calls disagreeing if `Value` is ever mutated in between.

File: cl/beacon/handler/block_production.go [hunk 12]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+	}
+	externalBid, found := a.epbsPool.HighestBids.Get(bidKey)
+	selectedValueWei, selected := selectHigherGloasBidValue(block.ExecutionValue, externalBid, boostFactor)
+	if !found || !selected {
+		blockMachine, err := processBlockForProduction(baseState, block)
+		return baseState, blockMachine, err
+	}
+
+	// ProcessBlock mutates state, so keep baseState untouched for a possible self-build fallback.
+	candidateState, err := baseState.Copy()
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go:1023]: This puts a full beacon-state deep copy plus a second `ProcessBlock` on the block-production critical path, inside the slot deadline.

When an external bid wins and then fails, the request pays: `baseState.Copy()`, `ProcessBlock` on the copy, `ProcessBlock` on `baseState`, and finally `HashSSZ` of the post state. On mainnet the copy alone is not cheap, and it happens on every slot where an external bid beats the local value.

Worth measuring before Gloas goes live; if it is too slow, the cheaper shape is to validate the bid against the state once (the bid-specific checks) rather than replaying the whole transition twice.

File: cl/beacon/handler/block_production.go [hunk 13]
```diff
@@ -945,6 +917,140 @@ func (a *ApiHandler) produceBlock(
...
+		log.Warn("GLOAS: failed to copy state for external bid; using self-build",
+			"slot", block.Slot,
+			"builderIndex", externalBid.Message.BuilderIndex,
+			"bidValueGwei", externalBid.Message.Value,
+			"err", err)
+		blockMachine, processErr := processBlockForProduction(baseState, block)
+		return baseState, blockMachine, processErr
+	}
+
+	selfBlobs := block.Blobs
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production.go]: The save/restore pattern mutates the caller's `block` in place and then puts it back by hand.

Four fields are cleared and four are restored, in two places that must stay in sync. Any field added later to `BlindOrExecutionBeaconBlock` that differs between a self-build and an external bid (another sidecar list, a cached envelope, a proof set) will be missed silently — the block will simply carry stale self-build data into the external-bid response.

A shallow copy of `block` with the external bid applied, discarded when the candidate transition fails, would make the restore unnecessary and the invariant structural instead of manual.

File: cl/pool/epbs_pool.go [hunk 14]
```diff
@@ -37,6 +39,8 @@ type HighestBidKey struct {
 // EpbsPool holds EPBS-related gossip data caches.
 // [New in Gloas:EIP7732]
 type EpbsPool struct {
+	highestBidsMu sync.Mutex
```
Comment (AskAlexSharov) [inline cl/pool/epbs_pool.go]: This mutex only guards two of the four access paths, so it reads as stronger protection than it gives.

`HighestBids` stays exported and is still read directly by `executionPayloadBidService.validateHighestBid`, by `ApiHandler.processProducedBlock`, and by the `epbs.go` highest-bid handler — none of them take `highestBidsMu`. The hashicorp LRU already has its own lock, so those reads are safe on their own; the only thing this mutex actually buys is the compare-and-delete inside `RemoveHighestBid`.

Either unexport `HighestBids` and route every access through the pool methods, or name the lock for what it protects (the CAS) so a later reader does not assume the whole cache is serialized here.

File: cl/beacon/handler/block_production_test.go [hunk 15]
```diff
@@ -703,6 +718,454 @@ func TestProduceBeaconBodyRejectsMissingBlobsBundleAtDeneb(t *testing.T) {
...
+
+	require.NoError(t, h.broadcastBlock(t.Context(), block))
+
+	// The persistence error is logged at the end of the background store goroutine.
+	require.Eventually(t, func() bool {
+		return strings.Contains(logs(), "stop after persistence")
+	}, 5*time.Second, 10*time.Millisecond)
+	require.Contains(t, logs(), "blobSidecars=0")
+	require.Contains(t, logs(), "columnSidecars=0")
+	require.NotContains(t, logs(), "blobs=1")
```
Comment (AskAlexSharov) [inline cl/beacon/handler/block_production_test.go]: This assertion can never fail.

The same PR renames the log key from `blobs` to `blobSidecars`/`columnSidecars`, so no line in `BlockPublishing: publishing block and sidecars` contains `blobs=` at all any more. The check passes for every possible outcome, including a regression that reintroduces local sidecar building.

`require.Contains(t, logs(), "blobSidecars=0")` on the line above already carries the real signal; this line can go, or be replaced by an assertion that no `missing blob bundle` error was returned.

File: cl/pool/epbs_pool_test.go [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package pool
```
Comment (AskAlexSharov) [inline cl/pool/epbs_pool_test.go]: New test file where an existing one would do: `cl/pool/operations_pool_test.go` is already in this package. The project convention is to append to the existing `_test.go` rather than add a file per source file.

File: cl/beacon/handler/block_production.go (hunk 4 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go]: Fixed in 62b8ca81d6. `ProcessExecutionPayloadBid` now wraps only deterministic bid-validation failures with `ErrInvalidExecutionPayloadBid`. The production fallback evicts the pooled bid only when `errors.Is(candidateErr, eth2.ErrInvalidExecutionPayloadBid)` is true; operational or otherwise unclassified errors, including state-access and signature-verification backend errors, keep the bid available for retry. I also added a regression test that injects a non-bid transition failure and verifies that the self-build succeeds while the external bid remains in the pool. Existing tests continue to verify that deterministically invalid bids are evicted.

File: cl/beacon/handler/block_production.go (hunk 5 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go]: Fixed in 9358163810. The ProduceBlockV3 query value is now explicitly limited to the legacy relay-versus-local path. `processProducedBlock` and the Gloas selector no longer accept that factor, so p2p bids are compared directly with the local Wei value at the default 100% weighting, with the local build winning ties. The regression test failed before the change when the legacy value was zero and now confirms that the higher p2p bid is selected. The complete spec-compliant ProduceBlockV4 `BuilderConfig` handler is tracked in #23686.

File: cl/beacon/handler/block_production.go (hunk 6 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go:954]: Fixed in 519f1f8816. `processProducedBlockWithProcessor` now obtains the self-build value through `block.GetExecutionValue()` before comparison, so nil is normalized to zero. The normalized value is also reused for logging and fallback restoration. Added `TestProcessProducedBlockTreatsNilExecutionValueAsZero`; it reproduced the nil dereference before the fix and now passes.

File: cl/pool/epbs_pool.go (hunk 7 above)
Comment (yperbasis) [reply cl/pool/epbs_pool.go:72]: Agreed. The current eviction preserves proposal liveness through the self-build fallback, but it can leave the market without a previously validated external candidate. Clearing `seenCache` would weaken duplicate-bid protection and would not recover other discarded bids, so the complete fix needs a bounded, value-ordered candidate set with atomic selection and eviction plus a bounded retry policy. Tracked in #23689.

File: cl/beacon/handler/block_production.go (hunk 8 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go:1043]: Fixed in 5defc8f2d4. `GetEthV3ValidatorBlock` now updates the slot cache only after production succeeds: it stores the envelope produced by a self-build, or removes any existing slot entry when the final result uses an external bid. Errors return before this cache update, so a failed retry does not discard a usable earlier envelope. Added `TestGetEthV3ValidatorBlockClearsStaleSelfBuildEnvelopeForExternalBid`, which performs two real V3 production calls for the same slot—a self-build followed by an external-bid selection—and verifies that the stale envelope is removed.

File: cl/beacon/handler/block_production.go (hunk 9 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go:969]: Good point about the ProduceBlockV4 valuation rule. This helper only selects bids admitted through the ePBS gossip pool, where a non-zero `execution_payment` is a protocol rejection and is rejected before pool insertion. Non-zero execution payments come from direct ProduceBlockV4 Builder API entries and must be valued using the applicable `max_execution_payment`; adding the payment here without that cap would be incorrect.

I renamed the helper to `selectHigherGloasP2PBidValue` in c6cb26071a to make this boundary explicit. The complete V4 path, including capped execution payments, is tracked in #23686.

File: cl/beacon/handler/block_production.go (hunk 10 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go:876]: Agreed. I took the second option in 4f92a7b8e6. `shouldRequestBuilderHeader` now owns only the pre-Gloas protocol gate, while `getBuilderPayload` remains the single source of truth for builder enablement and client availability. This makes `errBuilderNotEnabled` reachable for disabled pre-Gloas configurations, keeps that expected case silent at the caller, and still avoids the legacy Builder API request on Gloas.

File: cl/beacon/handler/block_production.go (hunk 11 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go]: Agreed. Fixed in d01ba01b95. `getBuilderPayload` now parses and validates `BlockValue()` once and returns the resulting `*big.Int` with the header. `produceBlock` reuses that exact value for comparison, logging, and response metadata, so the validated wire value is never parsed a second time. I kept the parsed value out of `ExecutionHeader` to avoid adding derived mutable state to the wire-format type.

File: cl/beacon/handler/block_production.go (hunk 13 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production.go]: Agreed. Fixed in b7e4cfc7d9. The external path now shallow-copies both `BlindOrExecutionBeaconBlock` and its pointer-held `BeaconBody`, applies the bid, value, and empty sidecars only to that candidate, and processes the candidate copy. A successful candidate replaces the caller block; a failed candidate is discarded, so the original self-build reaches fallback untouched and no restore list remains. The extra work is limited to two small shallow struct copies; nested blobs, proofs, operations, and lists are not copied, and the existing state copy and transition count are unchanged.

File: cl/pool/epbs_pool.go (hunk 14 above)
Comment (yperbasis) [reply cl/pool/epbs_pool.go]: Addressed in 98411f3763. `HighestBids` is now private, and all callers use `GetHighestBid`, `HighestBidKeys`, `StoreHighestBid`, or `RemoveHighestBid`. I also renamed the mutex to `highestBidUpdatesMu` and documented that the LRU synchronizes individual operations while this mutex makes replacement atomic with conditional removal. This prevents callers from bypassing the update/remove contract with raw cache mutations.

File: cl/beacon/handler/block_production_test.go (hunk 15 above)
Comment (yperbasis) [reply cl/beacon/handler/block_production_test.go]: Addressed in 6c42c29ba6. I removed the obsolete `blobs=1` negative assertion. The test already verifies that `broadcastBlock` succeeds without a local blob bundle and reports `blobSidecars=0` and `columnSidecars=0`, which directly cover the intended behavior.

File: cl/pool/epbs_pool_test.go (hunk 16 above)
Comment (yperbasis) [reply cl/pool/epbs_pool_test.go]: Addressed in 484bcfbcb9. I moved `TestRemoveHighestBidOnlyRemovesMatchingBid` into the existing `operations_pool_test.go` and removed the standalone `epbs_pool_test.go`. This is only a test-layout change; the test behavior is unchanged.

File: cl/beacon/builder/types.go [hunk 17]
```diff
@@ -48,10 +48,14 @@ func (h ExecutionHeader) BlockValue() *big.Int {
 	if h.Data.Message.Value == "" {
 		return nil
 	}
-	//blockValue := binary.LittleEndian.Uint64([]byte(h.Data.Message.Value))
 	blockValue, ok := new(big.Int).SetString(h.Data.Message.Value, 10)
 	if !ok {
 		log.Warn("cannot parse block value", "value", h.Data.Message.Value)
+		return nil
+	}
+	if blockValue.Sign() < 0 || blockValue.BitLen() > 256 {
```
[the remark this answers is not in the document]
Comment (yperbasis) [reply cl/beacon/builder/types.go]: Fixed in eda428ce72. `BlockValue` now requires the raw value to contain only ASCII decimal digits before calling `SetString`, so signed forms such as `+1` are rejected. Added a `leading_plus` case to `TestGetBuilderPayloadRejectsInvalidBlockValue`; it failed before the parser change and passes now.
